### PR TITLE
feat(charts): add chart display settings

### PR DESCRIPTION
**Pull request by Dhruwang**
## What & why

Charts had no saved display settings. The `Chart.config` JSON column and its `ZChartConfig` schema existed, but every save path wrote `{}`, so nothing in it was ever exercised. The ticket calls out two problems: open-text answers cannot be surfaced in a chart, and bar charts are always vertical.

**This PR solves the second one only.** It adds a **Chart display settings** section between the preview and the chart name, with one setting persisted in `Chart.config` and honored wherever the chart renders — preview, chart list, dashboard widget:

- **Bar direction** (bar charts only) — vertical vs. horizontal bars. `CartesianChart` gains a `horizontal` mode that swaps the axis roles (values on the x-axis, categories down the y-axis), flips the gridlines, wraps the category labels in a tick sized to the labels actually present, and moves value labels to the end of each bar.

Both axis measurements are pure functions in `chart-utils` — `getCategoryAxisWidth` (the left gutter, sized to the labels present) and `getValueLabelPadding` (the right gutter, sized to the widest value label) — so they carry unit tests rather than living inside the component.

The section hides itself entirely for chart types with no applicable settings, rather than rendering a heading with nothing under it.

Settings a chart type can't use are dropped on save (`sanitizeChartDisplay`) instead of stored as dead values, so switching a bar chart to a pie chart doesn't leave a stale orientation behind.

I deliberately did **not** add toggles for the other pre-declared `ZChartConfig` fields (`showLegend`, `showGrid`, `stacked`, `colors`, …). None of them are wired to the renderer today, so each would be a control that changes nothing — worse than not shipping it. The section is structured so they can be added one at a time as the renderer grows support.

### Open text is deliberately still open

A **Display chart / Display data table** toggle was built here first, which is what made the data table a real, savable rendering and so closed the open-text half of the ticket. It has been pulled back out: sitting next to the chart type picker, it read as two controls for one decision — pick "Bar chart", then "Display data table", and the chart type stops meaning anything.

The open question for @jobenjada is whether the table should instead be **its own entry in the chart type picker**, so a single control answers "how should this render?". That needs a one-line `ChartType` enum migration, so it is worth deciding before it ships rather than after. Until then this PR leaves the preview's Chart/Data tabs exactly as they are on main, and open text stays where it was — visible while building a chart, not persisted onto a dashboard.

`displayType` is out of `ZChartConfig`; zod strips it from any config already saved with it, so nothing needs migrating.

### The horizontal value-label gutter

The flipped value axis started as a bare `<XAxis type="number">` with no `padding`, while the vertical `<YAxis>` it replaces carries `padding={{ top: 16, bottom: 4 }}` — and that top padding is what keeps `LabelList position="top"` inside the SVG when a bar reaches the top of the domain. Horizontally the label moves to `position="right"`, past the bar's end, with nothing reserving room for it, so the longest bar lost its number whenever `dataMax` landed exactly on the axis bound — which `computeYAxis` produces routinely (`niceMax` is a multiple of the step, so 10 / 20 / 50 / 100 all hit it, as does a pinned rating scale at full marks).

Rendered both axis configurations to confirm it rather than eyeballing, at `width={400}`, `domain=[0,50]`, data `[50, 20, 5]`:

| | longest bar's value label, x | inside the 400px viewport? |
|---|---|---|
| no `padding` | **400** (`textAnchor="start"`) | no — anchored on the edge, runs outward |
| `padding={{ left: 4, right: 26 }}` | **379** | yes |

The gutter is sized from the labels actually present (`getValueLabelPadding`), not a fixed two-digit guess, and capped so a very large number can't eat the plot.

### Also in this branch

- **Options switch is a `fieldset`, not a `div role="group"`.** The container carries grouping semantics natively, so the label referenced by `aria-labelledby` names the whole switch. `min-w-0` undoes the browser's default `min-width: min-content` on `fieldset`, which would otherwise stop it shrinking in a flex parent. Found while building the Bar direction control, which uses this component.

### What arrived from the stacked PR

**#8899 (the pie chart's Breakdown bars display) has been merged into this branch**, so its diff is now part of this PR: `ZChartConfig` also gains `pieDisplay`, the display settings card gains a **Display as** setting for pie charts, and `breakdown-bars.tsx` renders a pie's data as one horizontal bar split by share. Both settings are resolved and sanitized by the same pair of helpers, so each chart type keeps only the setting it can use.

### Percentage shares are locale-aware

Both places that print a share — the pie's slice labels and the breakdown bar's tooltip/legend — used `toFixed(1)` with a literal `%`, which always renders a period decimal separator (`12.5%` where de-DE wants `12,5 %`). They now share one helper, `formatPercentShare(percent, locale)`, over `Intl.NumberFormat(… style: "percent" …)`. The single fraction digit is kept on purpose: whole percents print a real 0.4% group as `0%` and make three equal groups sum to 99%.

The pie's label renderer is curried with the locale — recharts calls it outside React, so the language is closed over rather than read from a hook inside it.

### Scope note after merging main

Merged `main` again after **ENG-2473 (#8921)** landed, which added the per-widget chart/data view toggle and touches the same three files this branch does (`chart-renderer`, `dashboard-widget-data`, `dashboard-detail-client`). Both sides were kept everywhere: a widget now receives both the chart's saved `config` and its `view`.

This branch previously also carried the **pie tooltip fix (ENG-2346)** and a revert of two local entitlement stubs. Both are now on `main` — the tooltip fix shipped separately as #8903 — so neither appears in this diff any more. Conflict resolution took main's version of that fix; the code was identical on both sides, only the comment wording differed.

One remnant is intentional: `chart-utils.test.ts` keeps a regression test asserting `prepareMeasureSliceData` sets `tooltipLabel`, which main's version of the fix did not include. It guards the same behaviour from the other side.

Screenshot 2026-08-18 at 4 01 08 PM

https://github.com/user-attachments/assets/a6188cef-8980-4786-9e81-0322b60866bf

## Linear ticket

Ref ENG-1550

## How this was tested

- `npx vitest run modules/ee/analysis` on the merged tree → 28 files / **458 tests** ✅
- `formatPercentShare` has its own cases: `0.4%` / `33.3%` / `100.0%` in en-US, and de-DE renders `12,5` rather than `12.5`.
- `chart-utils.test.ts` gains 5 cases for the two lifted axis helpers: the gutter shrinks to short numeric labels, holds its floor and ceiling, takes the longest label rather than the first, reserves room for the widest value label, and caps a huge number.
- New unit tests in `apps/web/modules/ee/analysis/charts/lib/chart-display.test.ts` (13 cases): empty/absent config resolves to vertical bars (the pre-existing behavior for already-saved charts), partial config fills only the missing half, only `bar` supports an orientation, and sanitize drops the orientation for non-bar types while preserving unrelated config fields.
- `npx tsc --noEmit -p tsconfig.typecheck.json` → **0 errors** across the app. (Earlier runs on this branch reported errors in `modules/storage`; those were a stale `dist/` for unbuilt workspace packages, not real — `pnpm build --filter=@formbricks/web^...` clears them.)
- `npx eslint` on all touched files ✅ · `npx prettier --check` on the touched areas ✅
- `pnpm i18n` ✅ — en-US strings added and all other locales regenerated (0 missing, 0 unused keys)
- Manual: screenshot and recording above show the setting in the builder and a horizontal bar chart rendering from it.

## Breaking changes

None — `ZChartConfig` gains one optional field (`barOrientation`), and an empty config (every chart saved so far) resolves to the current rendering.

## QA / Test Plan

**How to test**

- [ ] Analysis → Charts → **Create chart** → build any chart (e.g. group by Question, measure Rating: Average) → once the preview renders, a **Chart display settings** card appears between the preview and **Chart Name**.
- [ ] Pick chart type **Bar** → a **Bar direction** setting appears → choose **Horizontal bars** → bars run left-to-right, question labels sit in a column down the left side, value labels appear at the right end of each bar, gridlines run vertically.
- [ ] Save the horizontal bar chart, reopen it and view it on a dashboard → still horizontal in both places.
- [ ] Switch the chart type from **Bar** to **Pie** (or Area/Line/Big Number) → the **Bar direction** setting disappears; save, reopen, switch back to **Bar** → direction is back to **Vertical bars** (the unsupported setting was not kept).
- [ ] Existing charts created before this change → open one and view it on a dashboard → unchanged, with **Vertical bars** preselected.
- [ ] Edge case: a bar chart with only measures and no group-by (e.g. sentiment counts) → horizontal mode still renders one labelled bar per measure with no empty gap.
- [ ] Edge case: a bar chart with very long category labels → labels wrap in the left-hand tick instead of overlapping or being cut off.
- [ ] Edge case: a horizontal bar chart whose max is a round number (e.g. exactly 50 or 100 responses, or a rating pinned at full marks) → the longest bar still shows its value label; nothing is clipped at the right edge.
- [ ] Edge case: a horizontal bar chart with only small numeric categories (e.g. 1–5) → the left gutter is narrow, not a fixed wide column.
- [ ] Edge case: a query returning no rows → "No data returned" message, no crash.
- [ ] Add a widget to a dashboard and toggle its **chart / data** view (from #8921) → the saved bar direction or pie display still applies in the chart view, and the data view still shows the table.
- [ ] Switch the app language to German → a pie's slice labels and a breakdown bar's legend read `12,5 %`, not `12.5%`.
- [ ] Options switch a11y: with a screen reader on the **Bar direction** switch, the group is announced with its label; visually, the switch still shrinks correctly in a narrow panel rather than overflowing.

**Preconditions / test data**

- Dashboards + feedback directories entitlement enabled for the organization (EE feature).
- A feedback directory with records across several questions, including at least one long question label.
- At least one dashboard to add widgets to.

**Risks & regressions**

- Bar chart rendering generally — `CartesianChart` now branches its axes on the new `horizontal` prop. Vertical bar, area, line and pie charts must look exactly as before (axis labels, wrapping, fixed rating/CSAT/NPS scales, tooltips).
- Chart save/update and "add to dashboard" all now send a real `config`; confirm saving, editing and duplicating charts still work.
- The dashboard widget path reads the saved config too (`dashboard-detail-client`, `dashboard-widget-data`), so a widget must honour the orientation the chart was saved with.
- `OptionsSwitch` is shared — its container changed from a `div` to a `fieldset`. Check the other places it is used still lay out and shrink correctly.

**Migrations / env / cutover**

- none — `Chart.config` already exists and is a JSON column; the new field is optional.

### CODE REVIEW

**Review by mattinannt — COMMENTED** [review]
Reviewed the diff end to end (chart-display.ts/test, cartesian-chart.tsx, chart-renderer.tsx, chart-preview.tsx, data-viewer.tsx, chart-display-settings.tsx, use-chart-dialog.ts, dashboard widget wiring, and the locale/i18n changes).

**Overall:** solid, well-scoped change. The core logic in `chart-display.ts` (`resolveChartDisplay`/`sanitizeChartDisplay`/`supportsBarOrientation`) is clean and the unit tests cover the meaningful edge cases (empty/missing config, partial config, sanitize dropping unsupported settings while preserving unrelated fields). Traced the axis-flip math in `CartesianChart` (layout swap, `CartesianGrid` `vertical`/`horizontal` flip, `WrappingYAxisTick`) and it's internally consistent with the existing vertical-axis logic — the gridline flip (`vertical={horizontal} horizontal={!horizontal}`) is correct for Recharts' layout semantics. Config threads through consistently: preview → save (`sanitizeChartDisplay` using the freshly-resolved chart type, including on the "ensure chart for dashboard" auto-create path) → `ChartRenderer` (chart list, dashboard widget) all resolve through the same `resolveChartDisplay`, so there's no divergent behavior between where a chart renders. Confirmed the removed i18n keys (`charts.chart`, `charts.chart_data`, `charts.chart_data_tab`) have no remaining references, and the locale diffs are consistent 1:1 with the en-US additions.

Two minor, non-blocking notes:

- `chart-display-settings.tsx`: the `<Label htmlFor="chart-display-type">` / `<Label htmlFor="chart-bar-orientation">` point at a wrapping `<div>` around `OptionsSwitch`, not a focusable form control, so clicking the label won't do anything. Existing `OptionsSwitch` usages elsewhere (e.g. `edit-ending-card.tsx`, `multiple-choice-element-form.tsx`) just use a bare `<Label>` with no `htmlFor`, which avoids implying an association that doesn't do anything. Not a regression, just worth aligning with the existing pattern.
- The PR body flags that no screenshots/recording could be captured in this environment (needs a live Cube/Hub backend + feedback directory). Given this touches chart rendering directly (axis flip, label positioning, gridlines), it'd be good to get that visual confirmation — especially horizontal bars with long question labels and the multi-measure/no-dimension bar case — before merge, even if the logic checks out from reading the code.

Nothing here blocks approval from my read of the diff; CI (typecheck/lint/unit tests) is green and the remaining checks (E2E, SonarQube, Docker build) were still in progress at review time.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by itsjavi — CHANGES_REQUESTED** [review]
thanks for the PR, this touches chart rendering directly (axis flip, label wrapping), can you provide more screenshots or a recording?

**Comment by Dhruwang** [comment]
@itsjavi i now attached a screen recording ✅

**Review by itsjavi — DISMISSED** [review]
thanks for the PR and details! LGTM
I removed the breaking-changes label since it seemed a false positive

**Review by kiln-agent — COMMENTED** [review]
Reviewed the merged diff at `2c307d6` end to end: `chart-display.ts`/`.test.ts`, `cartesian-chart.tsx`, `chart-renderer.tsx`, `chart-preview.tsx`, `create-chart-view.tsx`, `use-chart-dialog.ts`, the dashboard wiring, `options-switch`, `ZChartConfig`, and the locale changes.

**Overall:** well-scoped and the plumbing is right. `resolveChartDisplay`/`sanitizeChartDisplay`/`supportsBarOrientation` are pure and covered by the new tests; every render path (preview, dashboard widget) resolves through the same `resolveChartDisplay`, so a chart can't render differently depending on where it's shown; all three save paths sanitize against the *freshly resolved* chart type (including the "ensure chart for dashboard" path); `selectChart` already includes `config`, so the widget really does get it; and `duplicateChart` round-trips it through `ZChartConfig.parse`. Absent config → vertical, so existing charts are untouched. Verified the removed-key worry from the earlier round is moot — no i18n keys are deleted in this diff.

I found one thing I think should be fixed before merge.

---

### 1. Horizontal bars: the longest bar loses its value label

`cartesian-chart.tsx:262` · `chart-renderer.tsx:193,233`

The flipped value axis is a bare `<XAxis type="number" …>` with no `padding`. The vertical `<YAxis>` it replaces carries `padding={{ top: 16, bottom: 4 }}` (`cartesian-chart.tsx:309`) — and that top padding is exactly what keeps `LabelList position="top"` inside the SVG when a bar reaches the top of the domain. In horizontal mode the label moves to `position="right"`, past the bar's end, with nothing reserving room for it.

That matters whenever `dataMax` equals the axis domain max, which `computeYAxis` produces routinely: `niceMax = ceil(spanHigh / step) * step`, so any max that is a multiple of the derived step lands exactly on the bound — 10, 20, 50, 100 all do — and a pinned rating/CSAT/CES/NPS scale does too when the average sits at full marks.

I rendered both axis configurations to check rather than eyeballing it. With `width={400}`, `domain=[0,50]`, `ticks=[0,10,…,50]` and data `[50, 20, 5]`:

| | longest bar ends at | its label anchored at | inside the 400×300 viewport? |
|---|---|---|---|
| horizontal (this PR) | `x = 395` | `x = 400`, `textAnchor="start"` | **no** — starts on the right edge and runs outward |
| vertical (main) | `y = 21` | `y = 16` | yes, thanks to `padding.top` |

`recharts`' `Surface` renders a plain `<svg width height viewBox>` and `ChartContainer` sets no `overflow: visible`, so the SVG viewport clips it — the number on the biggest bar, the one people read first, just isn't there. The 20 and 5 labels are fine (`x = 190`, `x = 85`), which is why this doesn't show in the recording: that chart's max doesn't hit its bound.

Mirroring the vertical axis fixes it, e.g. `padding={{ left: 4, right: 24 }}` on the horizontal `XAxis` — worth sizing the right value for the widest label `formatCellValue` can emit rather than a 2-digit count. Adding "a bar chart whose max is a round number (e.g. exactly 50 or 100 responses)" to the QA list would catch this class of thing.

### 2. Two comments describe an earlier version of the code

- `cartesian-chart.tsx:50` points at "`getCategoryAxisWidth`" — no such symbol exists anywhere in the repo; the logic is the inline `categoryAxisWidth` `useMemo`.
- `cartesian-chart.tsx:43` says the category labels "get a fixed-width column instead of the wrapping tick", which is the opposite of what shipped: the width is content-sized and the labels *do* use a wrapping tick (`WrappingYAxisTick`).

Both read as leftovers from an iteration where the gutter was a flat 160px. Since the `categoryAxisWidth` computation is pure logic, lifting it into `chart-utils` as the `getCategoryAxisWidth` the comment already assumes would make the prose true and give it a unit test without a `.tsx` test.

### 3. `OptionsSwitchProps` is not `Readonly<>`

`options-switch/index.tsx:27` — AGENTS.md is unambiguous ("Always mark React component props as `Readonly<>`"), and the diff already edits that destructuring line to add `aria-labelledby`. CodeRabbit raised this last round too.

### Minor observation, not a request

`WrappingYAxisTick` floors `boxHeight` at `X_AXIS_LABEL_LINE_HEIGHT` (16px), so once the band drops under ~24px (roughly 20+ categories in a standard-height widget) neighbouring label boxes overlap by a few px. The bars are equally thin at that density, so I don't think it's worth solving here — but the QA list checks long labels and not many categories, and that's the case where the clamp runs out.

### What I ran

- `npx vitest run modules/ee/analysis/charts/lib` → 90 tests pass. `charts.test.ts` fails locally on `Failed to resolve entry for package "@formbricks/database"`, which is the unbuilt-`dist/` trap documented in AGENTS.md, not this PR; the **Unit Tests** job on CI is green, as are typecheck, lint, E2E, SonarQube and the Docker build on `2c307d6`.
- Rendered the two axis configurations above under jsdom to confirm finding 1 and to confirm the vertical path is unaffected (throwaway, not committed).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Dhruwang** [comment]
All three addressed in d66ad9f.

**1. Clipped value label** — confirmed and fixed. The flipped axis now carries `padding={{ left: 4, right: valueLabelPadding }}`, sized from the labels actually present rather than a two-digit guess (`getValueLabelPadding`, capped at 72px so a very large number can't eat the plot). Rendered both configurations to check the fix, same setup you used — `width={400}`, `domain=[0,50]`, data `[50, 20, 5]`: the longest bar's label goes from `x = 400` (anchored on the viewport edge, `textAnchor="start"`) to `x = 379`, with the other two still at 196 and 104.5. The QA list now has the round-number-max case.

**2. Stale comments** — both fixed, and taken the suggestion: `getCategoryAxisWidth` now exists in `chart-utils` alongside `getValueLabelPadding`, so the prose is true and the arithmetic has 5 unit tests instead of living in a `.tsx`. The render check above stayed a throwaway rather than a committed component test.

**3. `Readonly<OptionsSwitchProps>`** — done.

`npx vitest run modules/ee/analysis` → 427 passed / 27 files. eslint, prettier and `tsc --noEmit` clean on the touched files.

**Review by itsjavi — DISMISSED** [review]
thanks for the PR, approving but a few things to have an eye on as a follow up, because would make the chart feel broken:

- Horizontal bars reserve up to 232px for the two gutters, with no idea how wide the container is. Worth checking a half-width widget on a small screen.
- Multi-measure bar charts hide their value labels, but still reserve the gutter for them. Up to 72px of plot lost with nothing in it.
- The `>= VALUE_LABEL_MIN_PADDING` assertion cannot fail, since the helper clamps to that floor. A concrete expected value would catch a regression.

File: apps/web/modules/ee/analysis/charts/components/breakdown-bars.tsx [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  getMeasureAxisLabel,
+  sortMeasureIdsForCategoryAxis,
+} from "@/modules/ee/analysis/lib/schema-definition";
+import type { TChartDataRow } from "@/modules/ee/analysis/types/analysis";
+import { Tooltip, TooltipContent, TooltipProvider, TooltipTrigger } from "@/modules/ui/components/tooltip";
+
+// One decimal, the precision renderPieLabel prints, so the two displays of one pie chart never
+// disagree on a share: whole percents turn a real 0.4% section into "0%" and make three equal
+// groups add up to 99%.
+const formatPercent = (percent: number): string => `${(percent * 100).toFixed(1)}%`;
```
[the remark this answers is not in the document]
Comment (Dhruwang) [reply apps/web/modules/ee/analysis/charts/components/breakdown-bars.tsx]: Fixed in 6763d78 — and taken further than the one call site, since the same literal `%` sat in `renderPieLabel`.

Both now go through one helper, `formatPercentShare(percent, locale)` in `chart-utils`, wrapping `Intl.NumberFormat(locale, { style: "percent", minimumFractionDigits: 1, maximumFractionDigits: 1 })`. The one fraction digit is deliberate: whole percents print a real 0.4% group as `0%` and make three equal groups add up to 99%, and the two displays of one pie chart have to agree to the digit.

`breakdown-bars` reads `i18n.language` from `useTranslation()`. The pie's slice label renderer is curried with the locale (`createPieLabelRenderer(i18n.language)`) because recharts calls it outside React, so the language cannot come from a hook inside it.

Tests sit on the helper rather than the components: `0.4%` / `33.3%` / `100.0%` in en-US, and de-DE renders `12,5` not `12.5`. `npx vitest run modules/ee/analysis` → 458 passed / 28 files.

**Review by itsjavi — APPROVED** [review]
